Add totalconnect zones as binary sensors

austinmroczek opened the pull request:
## Description:

Represent totalconnect "zones" as Home Assistant binary sensors.  

This will add a large number of entities.  It's 46 for my setup.

**Related issue (if applicable):** fixes https://github.com/craigjmidwinter/total-connect-client/issues/2

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable): https://github.com/home-assistant/home-assistant.io/pull/11160

## Example entry for `configuration.yaml` (if applicable):
no changes to config

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) https://github.com/home-assistant/home-assistant.io/pull/11160

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [n/a] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [n/a] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

austinmroczek committed:
Merge pull request #1 from home-assistant/dev

update to latest HA

austinmroczek committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into dev

austinmroczek committed:
Bump skybellpy to 0.4.0

austinmroczek committed:
Bump skybellpy to 0.4.0 in requirements_all.txt

austinmroczek committed:
Added extra states for STATE_ALARM_TRIGGERED to allow users to know if
it is a burglar or fire or carbon monoxide so automations can take
appropriate actions.  Updated TotalConnect component to handle these new
states.

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant.git into dev

austinmroczek committed:
Merge pull request #2 from home-assistant/dev

update

austinmroczek committed:
Fix const import

austinmroczek committed:
Fix const import

austinmroczek committed:
Fix const imports

austinmroczek committed:
Bump total-connect-client to 0.26.

austinmroczek committed:
Catch details of alarm trigger in state attributes.

Also bumps total_connect_client to 0.27.

austinmroczek committed:
Merge branch 'dev' of
https://github.com/austinmroczek/home-assistant.git into dev

austinmroczek committed:
Change state_attributes() to device_state_attributes()

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant into dev

austinmroczek committed:
Move totalconnect component toward being a multi-platform integration.  Bump total_connect_client to 0.28.

austinmroczek committed:
Merge pull request #4 from home-assistant/dev

pull in latest

austinmroczek committed:
add missing total-connect alarm state mappings

austinmroczek committed:
Made recommended changes of MartinHjelmare at
https://github.com/home-assistant/home-assistant/pull/24427

austinmroczek committed:
Update __init__.py

austinmroczek committed:
Updates per MartinHjelmare comments

austinmroczek committed:
flake8/pydocstyle fixes

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant.git into dev

austinmroczek committed:
removed . at end of log message

austinmroczek committed:
added blank line between logging and voluptuous

austinmroczek committed:
more fixes

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Adding totalconnect zones as HA binary_sensors

austinmroczek committed:
fix manifest.json

austinmroczek committed:
flake8/pydocstyle fixes.  Added codeowner.

springstan commented:
@austinmroczek please format your code with [black](https://github.com/psf/black), e.g. `black {source_file_or_directory}`, use the style guiding tool [flake8](http://flake8.pycqa.org/en/latest/) e.g. `flake8 path/to/code/` and update the `CODEOWNERS` file by running `python3 -m script.hassfest` :)

austinmroczek committed:
Update formatting per @springstan guidance.

springstan commented:
Please solve this last Pylint error: https://dev.azure.com/home-assistant/Home%20Assistant/_build/results?buildId=16497&view=logs&j=8b66295a-05e1-52cf-5640-7a4f4e1975e2&t=6c4465fe-285a-5650-787c-2e3dbe1da192&l=13

austinmroczek committed:
Fixed pylint

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Add zone ID to log message for easier troubleshooting

austinmroczek committed:
Account for bypassed zones in update()

springstan commented:
@austinmroczek you will need to sort all imports by using [isort](https://github.com/timothycrosley/isort) to solve the lint error :)

austinmroczek committed:
More status handling fixes.

austinmroczek committed:
Fixed flake8 error

austinmroczek committed:
Another attempt at black/isort fixes.

austinmroczek commented:
@springstan please help.  I'm running into a problem with black and isort wanting to make different formatting changes to the same file.  I can't seem to make both happy, and it causes CI to fail.  Which should have precedence?  Sorry if I'm missing something, but I don't see anything helpful in  https://developers.home-assistant.io/docs/en/development_guidelines.html 

If I first run black on binary_sensor.py and run isort second, then CI fails for black.  If I first run isort and second black, then CI fails for isort.

springstan commented:
Just format it with black for now, push it and I will take a look ✌

austinmroczek commented:
I have PR # https://github.com/craigjmidwinter/total-connect-client/pull/53 in for total-connect-client which fixes a minor error there, and would let us remove the imports for binary_sensor.py.  Once that hits I'll make one more commit to account for that.  Side benefit is it should help with the formatting issue.

austinmroczek committed:
Bump total-connect-client to 0.50.  Simplify code using new functions in
total-connect-client package instead of importing constants.  Run black
and isort.

austinmroczek committed:
Fix manifest file

austinmroczek committed:
Another manifest fix

austinmroczek committed:
one more manifest fix

austinmroczek committed:
more manifest changes.

austinmroczek commented:
@springstan I think I solved the formatting issues, but now I can't figure out what I've done to the manifest file that I can't resolve this conflict....please help.

springstan commented:
@austinmroczek you will have to solve the merge conflict by accepting one solution e.g. your solution in this case.

austinmroczek committed:
sync up

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant into dev

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,104 @@
+                self._zone.zone_type_id,
+                self._zone_id,
+            )
+        return None
```
Indent this line once to solve the pylint error ✌

austinmroczek replied:
Fixing the indent produces a new pylint error.

springstan replied:
I think pylint prefers the following format:
```python
if ... :  
    return ...
if ... :
    return ...
if ... :
    return ...
_LOGGER.info(...)
return None
```
Hopefully this solves your pylint issues 😅

austinmroczek committed:
fix indent

austinmroczek committed:
one more pylint fix

austinmroczek committed:
Hopefully the last pylint fix

austinmroczek commented:
@springstan I think I have finally crossed the finish line.

springstan commented:
Great to hear! I will take a look at it tomorrow to give you some feedback.

springstan commented on homeassistant/components/totalconnect/alarm_control_panel.py:
```diff
@@ -32,10 +32,8 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-        name = location.get("LocationName")
-        alarms.append(TotalConnectAlarm(name, location_id, client))
+    for k in client.locations:
+        alarms.append(TotalConnectAlarm(client.locations[k].location_name, k, client))
```
Why do you need to change this to pass the location object (k) onto TotalConnectAlarm?

austinmroczek replied:
client.locations is a dictionary.  k is the key, which is the Location ID.   We need the Location ID for pretty much every function call for total-connect-client.  Sorry I could have used better variable names.  Want me to change that?

austinmroczek replied:
Also, the change was partially due to changes we made in total-connect-client where Location ID went from being an entry in the dictionary to being the key.

springstan replied:
> Sorry I could have used better variable names. Want me to change that?

Yes please change all variables names that are not self-explanatory.

> Also, the change was partially due to changes we made in total-connect-client where Location ID went from being an entry in the dictionary to being the key.

Okay thanks for clarifying.

springstan replied:
Please rename `k` here as well.

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+
+    client_locations = hass.data[TOTALCONNECT_DOMAIN].client.locations
+
+    for k in client_locations:
```
Rename `k` here.

austinmroczek replied:
commit  6f0e779

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attributes = {}
```
Instead of declaring an empty dictionary and then filling it please do it in one, e.g.
```python
dict = {
    "key": "value",
    "key": "value",
    ...
}
```

austinmroczek replied:
fixed

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+        """Return the state attributes."""
+        attributes = {}
+        attributes["zone_id"] = self._zone_id
+        attributes["zone_description"] = self._name
```
Which functionality does it bring to add another attribute that contains the same value as the name of this sensor?

austinmroczek replied:
Good point.  I think in early testing I was just printing out everything I could from total-connect-client and didn't stop to think.  Removed.

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+        )
+        self._unique_id = "TC {} zone {}".format(
+            locations[location_id].location_name, zone_id
+        )
```
Why are name and unique_id the same?

austinmroczek replied:
I changed it so they are now different from each other.  I also fixed unique_id so it is more likely to remain unchanged (the location_name is user changeable from within total-connect-client / Honeywell system).

austinmroczek committed:
make variable names understandable

austinmroczek committed:
create and fill dict in one step

austinmroczek committed:
Fix name and attributes

austinmroczek committed:
rename to logical variable in alarm_control_panel

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,101 @@
+        self._zone_id = zone_id
+        self._location_id = location_id
+        self._zone = locations[location_id].zones[zone_id]
+        self._name = self._zone.description
```
How long will the name of the binary_sensor be?

austinmroczek replied:
I can't find any documentation on a limit in the Total Connect API other than "it's a string."  The names are generally entered by hand at the physical alarm panel, so I don't think we'll find any that are really long.  Some of my names are cut off at 32 characters.  That may be a limitation of my alarm panel, which is an older model, but I don't know for sure.

springstan replied:
Okay thanks for providing more information about the description.

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,101 @@
+    client_locations = hass.data[TOTALCONNECT_DOMAIN].client.locations
+
+    for location_id in client_locations:
+        for zone in client_locations[location_id].zones:
```
Since you are using the zone_id and passing it to TotalConnectBinarySensor.
```suggestion
        for zone_id in client_locations[location_id].zones:
```

austinmroczek replied:
concur

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,101 @@
+        """Initialize the TotalConnect status."""
+        self._zone_id = zone_id
+        self._location_id = location_id
+        self._zone = locations[location_id].zones[zone_id]
```
Please get the zone before you create an instance of this class and pass it into the constructor as a parameter.

austinmroczek replied:
concur

springstan commented on homeassistant/components/totalconnect/alarm_control_panel.py:
```diff
@@ -32,10 +32,12 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    for location_id in client.locations:
+        alarms.append(
+            TotalConnectAlarm(
+                client.locations[location_id].location_name, location_id, client
```
Please create a local variable for the location_name beforehand and use it here to make it more readable.

austinmroczek replied:
concur

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,101 @@
+        self.update()
+
+    @property
+    def unique_id(self) -> str:
```
```suggestion
    def unique_id(self)
```

austinmroczek replied:
concur

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,101 @@
+class TotalConnectBinarySensor(BinarySensorDevice):
+    """Represent an TotalConnect zone."""
+
+    def __init__(self, zone_id, location_id, locations):
```
`self._is_tampered`, `self._is_on` and `self._is_low_battery` are missing in the constructor, set them to `None`

austinmroczek replied:
concur

austinmroczek committed:
Remove location_name from alarm_control_panel attributes since it is
already the name of the alarm.

austinmroczek commented on homeassistant/components/totalconnect/alarm_control_panel.py:
```diff
@@ -72,35 +74,35 @@ def device_state_attributes(self):
+        status = self._client.get_armed_status(self._location_id)
         attr = {
-            "location_name": self._name,
+            "location_name": self._client.locations[self._location_id].location_name,
```
This is actually a duplication of the name, so I just removed it instead.

austinmroczek committed:
Multiple fixes to improve code per @springstan suggestions

austinmroczek commented:
I addressed all of your comments and changes, but I made changes locally instead of using the Github web interface.  I think that's why it's still complaining you asked for changes.

austinmroczek reviewed (commented):
I think this closes out the "review."  Sorry still fairly new to git and Github.

springstan commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+            return DEVICE_CLASS_SMOKE
+        if self._zone.is_type_carbon_monoxide():
+            return DEVICE_CLASS_GAS
+        _LOGGER.info(
```
```suggestion
        _LOGGER.warning(
```

austinmroczek committed:
Update homeassistant/components/totalconnect/binary_sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

springstan commented:
@MartinHjelmare could you please take a look at this to make sure I am not missing something? Thanks 👍

MartinHjelmare commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+        self._location_id = location_id
+        self._zone = zone
+        self._name = self._zone.description
+        self._unique_id = "TC Location {} zone {}".format(location_id, zone_id)
```
We prefer f-strings over str.format.

austinmroczek replied:
I dropped the words. 

f-strings:  sorry I hadn't noticed the update to the style guide.  changed.

MartinHjelmare commented on homeassistant/components/totalconnect/alarm_control_panel.py:
```diff
@@ -32,10 +32,9 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-        location_id = location.get("LocationID")
-        name = location.get("LocationName")
-        alarms.append(TotalConnectAlarm(name, location_id, client))
+    for location_id in client.locations:
```
```py
for location_id, location in client_locations.items():
    location_name = location.location_name
    ...
```

austinmroczek replied:
changed

austinmroczek commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+    @property
+    def device_class(self):
+        """Return the class of this device, from component DEVICE_CLASSES."""
+        if self._zone.is_type_security():
```
It is a cached review.

austinmroczek commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+            return DEVICE_CLASS_SMOKE
+        if self._zone.is_type_carbon_monoxide():
+            return DEVICE_CLASS_GAS
+        _LOGGER.warning(
```
Removed the logging.

austinmroczek commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+        return self._name
+
+    @property
+    def state(self):
```
Sorry I think I misunderstood "is_on".  I originally thought "is_on" meant the device is powered/active and attempting to sense something, which I thought was separate from the state it sensed.  

I now think that "is_on" is when the device senses the on/true state.  Am I correct now.  If so, I will remove the code as suggested, but will also need to alter my is_on logic.

MartinHjelmare replied:
Yes, `is_on` is the boolean representation of the binary entity state.

austinmroczek replied:
Thanks for the clarification.  fixed `is_on`

austinmroczek commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+        self._is_on = None
+        self._is_tampered = None
+        self._is_low_battery = None
+        self.update()
```
changed

austinmroczek committed:
Multiple changes per @MartinHjelmare review

austinmroczek commented on homeassistant/components/totalconnect/alarm_control_panel.py:
```diff
@@ -72,35 +71,34 @@ def device_state_attributes(self):
-        status = self._client.get_armed_status(self._name)
+        status = self._client.get_armed_status(self._location_id)
         attr = {
-            "location_name": self._name,
```
undid the change

austinmroczek committed:
simplify alarm adding

MartinHjelmare commented on homeassistant/components/totalconnect/binary_sensor.py:
```diff
@@ -0,0 +1,96 @@
+        self._location_id = location_id
+        self._zone = zone
+        self._name = self._zone.description
+        self._unique_id = f"TC {location_id} {zone_id}"
```
We can remove TC too.

austinmroczek replied:
removed TC

austinmroczek committed:
Fix binary_sensor.py is_on

MartinHjelmare reviewed (approved):
Looks good!
